### ISSUE TITLE
ambiguous semantics of duty constraints

### ISSUE THREAD

**Issue by simonstey** [w3c/poe#211]
reopening as https://github.com/w3c/poe/issues/112#issuecomment-290617968 and issues outlined in #191 are still not addressed..

## Issue

http://w3c.github.io/poe/model/#rule:
> For a **Rule** to become effective, it MUST be valid (i.e., all it's Constraints MUST be satisfied).

http://w3c.github.io/poe/model/#constraint:
>  For a **Rule**, if the Constraint is satisfied then the action becomes effective for the enclosing **Rule**. 
> [..] A **Rule** becomes effective if all of its Constraints are satisfied.

http://w3c.github.io/poe/model/#constraint-class:
> If the expression evaluates to **true** the Constraint is satisfied.

[effective (ɪˈfɛk tɪv) adj.](http://www.thefreedictionary.com/effective)
> 1. adequate to accomplish a purpose; producing the intended or expected result: effective teaching methods.
> 2. **in operation or in force; functioning; operative**: The law becomes effective at midnight.
> 3. producing a deep or vivid impression; striking: an effective photograph.
> 4. prepared and available for service, esp. military service. 

### Interpretation
```
1. A Rule becomes effective if all of its Constraints are satisfied
2. Duty is a type of Rule
3. in effect means in force/active 
4. A Constraint is satisfied if it evaluates to true
--------------
=> A Duty is only in force/active if all of its Constraints evaluate to true.
```
### Problematic Examples
```turtle
<http://example.com/policy:42>
    a odrl:Agreement ;
    odrl:obligation [ 
	# a odrl:Duty ;
        odrl:assignee <http://example.com/person:44> ;
        odrl:assigner <http://example.com/org:43> ;
        odrl:action odrl:compensate ;
        odrl:constraint [ 
            odrl:leftOperand odrl:payAmount ;
            odrl:operator odrl:eq ;
            odrl:rightOperand "500.00" ;
            odrl:unit <http://dbpedia.org/resource/Euro> ;   
        ]
    ] .
```
**should** => "person44 has the obligation to pay 500 euro to org43"
**is** => "As long as payAmount != 500 euro, the duty is not in effect hence no one has to pay anything"

---------------------
```turtle
<http://example.com/policy:42>
    a odrl:Agreement ;
    odrl:obligation [ 
	# a odrl:Duty ;
        odrl:assignee <http://example.com/person:44> ;
        odrl:assigner <http://example.com/org:43> ;
        odrl:action odrl:compensate ;
        odrl:constraint [ 
            odrl:leftOperand odrl:payAmount ;
            odrl:operator odrl:eq ;
            odrl:rightOperand "500.00" ;
            odrl:unit <http://dbpedia.org/resource/Euro> ;   
        ] ;
        odrl:constraint [ 
            odrl:leftOperand odrl:dateTime;
            odrl:operator odrl:gt ;
            odrl:rightOperand "2017-12-31" ;
        ]
    ] .
```
**should** => "Starting from 2018-01-01 onwards, person44 has the obligation to compensate org43. This duty is only considered to be fulfilled, if payed amount equals 500 euro"
**is** => "As long as payAmount != 500 euro or datetime <= 2017-12-31, the obligation to compensate org43 is not in effect"

--------
### Proposal

Distinguish between different interpretations of duty constraints:

`:duty odrl:constraint :c1` -> `:duty` is **in effect** iff `:c1` is satisfied
`:duty odrl:scope :c2` (or any other name for scope) -> `:duty` is **fulfilled** iff `:c2` is satisfied 

```turtle
<http://example.com/policy:42>
    a odrl:Agreement ;
    odrl:obligation [ 
	# a odrl:Duty ;
        odrl:assignee <http://example.com/person:44> ;
        odrl:assigner <http://example.com/org:43> ;
        odrl:action odrl:compensate ;
        odrl:scope [ 
            odrl:leftOperand odrl:payAmount ;
            odrl:operator odrl:eq ;
            odrl:rightOperand "500.00" ;
            odrl:unit <http://dbpedia.org/resource/Euro> ;   
        ] ;
        odrl:constraint [ 
            odrl:leftOperand odrl:dateTime;
            odrl:operator odrl:gt ;
            odrl:rightOperand "2017-12-31" ;
        ]
    ] .
```
**should** => "Starting from 2018-01-01 onwards, person44 has the obligation to compensate org43. This duty is only considered to be fulfilled, if payed amount equals 500 euro"
**is** => "Starting from 2018-01-01 onwards, person44 has the obligation to compensate org43. This duty is only considered to be fulfilled, if payed amount equals 500 euro"

**Comment by riannella** [comment]
Firstly, we clearly need to update the Rule to say:

> For a Rule to become effective, it MUST be valid (i.e., all it's Constraints MUST be satisfied **and any applicable duty relationships MUST be fulfilled**).


Note: we should add these **key** terms to the Terminology list as well:
- effective
- fulfilled
- satisfied

**Comment by riannella** [comment]
In the first Problematic Example it says:

> is => "As long as payAmount != 500 euro, the duty is not in effect hence no one has to pay anything"

I am not clear why _"...hence no one has to pay anything"_ was added?

person44 has to pay EU500, if not, the Constraint is evaluated to false, hence the Rule is not in effect - hence person44 has not meet his/her obligation.

**Comment by riannella** [comment]
I think the semantics for the second example is better matched if you use andSequence because the "should=>" is saying that the time constraint is first, followed by the payAmount constraint.

```
<http://example.com/policy:42>
    a odrl:Agreement ;
    odrl:obligation [ 
	# a odrl:Duty ;
        odrl:assignee <http://example.com/person:44> ;
        odrl:assigner <http://example.com/org:43> ;
        odrl:action odrl:compensate ;
        odrl:constraint [ 
            odrl:leftOperand cTime1 ;
            odrl:operator odrl:andSequence ;
            odrl:rightOperand cPay1 ;
        ]
    ] .
	
	cPay1:  a odrl:Constraint ;
	        odrl:leftOperand odrl:payAmount ;
            odrl:operator odrl:eq ;
            odrl:rightOperand "500.00" ;
            odrl:unit <http://dbpedia.org/resource/Euro> .
	
	cTime1:  a odrl:Constraint ;
	        odrl:leftOperand odrl:payAmount ;
            odrl:operator odrl:eq ;
            odrl:rightOperand "500.00" ;
            odrl:unit <http://dbpedia.org/resource/Euro> .
	
```

**Comment by riannella** [comment]
Perhaps the proposal should be:

> :obligation odrl:constraint :c1 -> :obligation is in effect iff :c1 is satisfied
> :duty odrl:constraint :c2  -> :duty is fulfilled iff :c2 is satisfied
>

**Comment by riannella** [comment]
I have updated the IM to reflect the difference in how a Duty behaves (ads a Rule) when it is subordinate to a Perm or Prohib (as opposed to be used as an obligation)

**Comment by simonstey** [comment]
that issue is still not fixed though.. 

https://w3c.github.io/poe/model/#duty-policy
> **If the Duty rule has been fulfilled** (that is, all its Constraints are satisfied), **then the Duty rule is in effect**. That is, the obligation has been met.

if a duty would only become effective if it has been fulfilled, why would someone fulfill it in the first place? (cf. https://github.com/w3c/poe/issues/211#issue-243981567)

if Alice's obligation to pay 5 € to Bob for downloading his song only becomes effective once she's paid him those 5 €, she could just download it for free, because the obligation wasn't [effective (ɪˈfɛk tɪv) adj.](http://www.thefreedictionary.com/effective).

**Comment by riannella** [comment]
Hmm...that should be better expressed as:

> A Policy MAY include an obligation to fulfill a Duty. If the Duty has been fulfilled (that is, all its Constraints are satisfied), then the obligation rule is in effect. That is, the obligation has been met.

I think the example:

> if Alice's obligation to pay 5 € to Bob for downloading his song only becomes effective once she's paid him those 5 €, she could just download it for free, because the obligation wasn't effective

That is a permission with a duty (not an obligation).

**Comment by nitmws** [comment]
(pre-note: @riannella and I have been creating a comment at almost the same time)

This looks like a terminology issue, I pointed at it in #215, there I wrote:
* ... by my understanding: "a Duty is in effect" = the action of the Duty has to be taken, "a Duty has been fulfilled" = the action of the Duty has been taken successfully.
* Pondering that again raises the issue: "(not) in effect" for Permissions and Prohibitions is used to tell "all constraints of a Permission or Prohibition have been satisfied, therefore the assignee(s) may do what has been permitted or must not do what has been prohibited" - BUT: can a Duty be "in effect" at all as if the constraints of a Duty Rule have been satisfied the action required by the Duty has been taken - and the related Permission or Prohibition may enter the "in effect" status by that.

The key problem is: in the logical context of a Policy a Duty is more a kind of constraint which has to be satisfied than a Rule. This is expressed by the position of a Duty in a Policy: a Duty MUST be related to a Permission or a Prohibition, it is not a "direct child" of a Policy. In other words: not the used class should define the terminology but the role of the property (using this class) in the information model of a Policy.

Therefore I think we have to adjust the terminology:
* **only Permissions and Prohibitions are "(not) in effect"**
* **all properties of type Duty - obligation, duty and consequence - should use "(not) fulfilled"** (using fulfilled to make a distinction to the explicit Constraints)

** Note on @riannella 's posting above: sorry, I see not semantic difference between obligation and duty. Both have to be fulfilled else something gets stuck with using the Rules of the Policy.

**Comment by riannella** [comment]
What I am proposing is that all Duties are fulfilled (like stated above), and for the case of an obligation, it **also** means the obligation is in effect - this makes it consistent for handling all policy-level rules.

**Comment by riannella** [comment]
@nitmws there is a significant difference between obligation and duty. The latter makes a permission effective or not (the former does not).

**Comment by riannella** [comment]
I think the core issue is that we are trying to force all Rules to "conform" to a specific concept which we call "effective" (or "in effect").

This, I believe, leads to mis-interpretations, such as that outlined in @simonstey's first comment (above). Because "effective" is interpreted as "in force", then if you don't pay the EU500, then the Duty is not "in effect", hence, it is not "in force", which has been interpreted as "i don't have to do the Duty now".

Since each Rule has subtle, but significant, differences, we should treat each independently when we are trying to evaluate the rule (so we no longer say "rules are effective..." but have specific terms for each rule type, as well as different outcomes/states for different use cases, which will also aid in the evaluator processing).

PROPOSAL:

1) Permission - The ability to perform an Action over an Asset

Starting state: not-allowed

If the constraints are satisfied, and duties fulfilled, then the permission action is "allowed" (for the assigned parties) over the Asset.

If the constraints are not satisfied, and/or duties not fulfilled, and the action has been exercised, then the permission action is "infringed" (for the assigned parties) over the Asset.

If the Permission is "infringed", and the constraints are satisfied, and duties fulfilled, and consequence fulfilled, then the permission action is "allowed" (for the assigned parties) over the Asset.


2) Prohibition - The inability to perform an Action over an Asset

Starting state: not-confirmed

If the constraints are satisfied, and the action has not been performed, then the prohibition action is "confirmed" (for the assigned parties) over the Asset.

If the action has been exercised, then the prohibition action is "violated" (for the assigned parties) over the Asset.

If the Prohibition is "violated", and the constraints are satisfied, and the remedy has been fulfilled, then the prohibition action is "confirmed" (for the assigned parties) over the Asset.


3) Duty - The obligation to perform an Action

Starting state: not-fulfilled

If the constraints are satisfied, and the action has been performed, then the duty action has been "fulfilled" (for the assigned parties).

If the constraints are not satisfied, and the action has not been exercised, then the duty action is "infringed" (for the assigned parties) over the Asset.

If the Duty is "infringed", and the constraints are satisfied, and the action has been exercised, and the consequence fulfilled, then the duty action has been "fulfilled" (for the assigned parties).


_Note: other, more appropriate, terms for "allowed", "infringed", "confirmed", "violated", "fulfilled" may be preferred.
Note: I hope I have captured all main cases, but may not have!_


SUMMARY:

- remove "effective" as it is misleading to use in all cases
- include the above proposal in the Rule class section (as a summary)
- update each subsection with the new narrative.

**Comment by riannella** [comment]
We can also show these in a table:
[states.pdf](https://github.com/w3c/poe/files/1240662/states.pdf)

**Comment by nitmws** [comment]
I point at my suggested solution in https://github.com/w3c/poe/issues/215#issuecomment-323785168 
and try to harmonise it with @riannella suggestion above:

* a key problem is that the current Duty Class a/ inherits a lot of design and processing rules from the Rule Class and not all the inherited things fit well into the Duty Class uses and b/ the Duty Class specification sets definition which don't harmonise with the different properties of type Duty Class (obligation, duty, consequence, remedy). See the list of usages of Duty Class by properties in my comment.
Suggestions: 
    * we should at least remove generic definitions from the Duty Class if they don't harmonise with the properties and define these details for each property individually
    * a larger step is to remove the Rule Class inheritance by making the Duty Class a stand-alone class.
* Re "fulfilled" and "infringed":
  * by the status design of Renato a Duty Class may have three stati
    * fulfilled: all what is required by a Duty Class instance is ok 
    * not fulfilled: at least one particle of a Duty Class instance is not ok
    * infringed: the first level Duty as such is not ok, the first level Duty has a consequence Duty and this Duty should be performed. 
  * As a consequence Duty exists only for Duty Classes used by a duty property in a Permission "infringed" is not generic. The processing model must be: check first the action and the constraints of the duty Duty. If the result is not ok the check if this duty Duty has consequence(s). If it has none then the result of the duty Duty is "not fulfilled", if it has one to many all of the consequence Duties must be performed. Only if all the consequence Duties return a "fulfilled" status the duty Duty is "fulfilled", else it is "not fulfilled".
  * My text above uses the terminology "... Duty is ok". This is an interim and internal status of a processing model of a Duty and may either be transformed into the formal "fulfilled" or "not fulfilled" or it may trigger the processing of consequences ...
  * ... but by my understanding also the "infringed" status is only an internal status, it will never be delivered as final status of a duty Duty instance. So I see "is not ok" and "infringed" in parallel: "infringed" exists only as "not ok" status for the duty Duty, while a "not ok" status of other kinds of Duty usages in obligation, consequence and remedy has not special name. If we need an internal status after checking the action and the constraints of a Duty Class instance and even if it only for the duty Duty we should have a single name for that - "infringed" is too harsh for that status in all cases, but that's only my personal view.
  * Renato's discussion of "infringed" raises the question: how to deal with violated Prohibitions? The definition of the use of remedy in 2.5.7 of the IM tells a Prohibition may be "violated". I suggest to use the same term for a "not ok duty Duty" and a "Prohibition action which has been taken": infringed, violated ...

My general suggestion: we need strict but this way also easy to follow (and as I hope understand) processing models for the use of a Duty class with each of the different properties and less generic specification in the Duty Class section.

**Comment by riannella** [comment]
1 - what "doesn't harmonise" specifically in the Rule properties and the additional Duty properties?

2 - I think the "state table" clears up some issues. If you look at "infringed", then this could happen anytime during the evaluation of the permission rule - so it actually becomes a public state that odrl processors need to identify ?
(actually the table should also have "exists" for consequence, for that row).

3 - this raise yet-another-state: "error"
For example, I have a prohibition to print (and no remedy defined) and I print it?

**Comment by nitmws** [comment]
re @riannella 

re 1: I may have to clarify: I think that the properties obligation, duty, consequence and remedy don't make use of the Duty Class in the same way. This not about the properties of the Duty or the Rule Class.
* duty: the "original" property. If it is in effect processing it must return a fulfilled/not-fulfilled status. While being processed it may have an infringed/violated internal status. The statuses of a duty have an impact on the permission Rule including it.
* consequence: same as duty regarding "(not) in effect" and the returned  status. Both have an impact on the duty Duty including it. It may have an internal infringed/violated status but this has no impact on anything. 
And: a consequence Duty must not have a consequence property.
* remedy: same as duty regarding "(not) in effect" and the returned status. Both have an impact on the prohibition Rule including it. Same as consequence regarding infringed/violated.
And: a remedy Duty must not have a consequence property.
* obligation: same as duty regarding "(not) in effect". A generated fulfilled/not-fulfilled status has no pre-defined impact on anything inside the Policy including obligation(s). (I guess this could be a use case: Policy 01 includes obligation(s) and a Permission (of another Policy ?) has this duty "all obligations of Policy 01 must be fulfilled"). An obligation may have an internal infringed/violated status but this has no impact on anything. 
And: does a target property make sense for an obligation?

Conclusion: 
* the use of the consequence property depends on the property of type Duty. A strict design would have a basic Class without a consequence property for the consequence and remedy properties plus a sub-class with consequence property for duty and obligation.
* The handling of an internal infringed/violated status is inconsistent for the different properties of type Duty.

re 2: I had a deeper look at the table - and suggest to rename it to rule-processing-statuses to clarify its message
* we should make readers aware that more than 1 constraints, duties, consequences or remedies may be in a Permission - and that they are ANDed.
* re "infringed": it is a surprised to see "infringed" as status of Permission. So far we have discussed only to name the processing result "the evaluation of taking the action and of the constraints of a Duty is not ok" as "infringed". And this is Duty-internal only: if there is no consequence this results in a "not fulfilled" Duty status, if there is a consequence it must be processed.
* Duty row(s):
  * I think a row intentionally without a status in the "action" column doesn't make sense. Or is only the "Not-Exercised" missing? 
  * For a full coverage of stati the case Infringed/Satisfied/ / Fulfilled/ /Not-Exercised/Not-Fulfilled should be included.
  * (I omit comments on Permission and Prohibition here, but see re 3 below)

Conclusion: such a table should be included into the IM - we need a clear and detailed processing model.

re 3: I recall the question "what happens in case of violated prohibitions without a remedy" was raised at a call. The response was "nothing, if something should happen the assigner has to define a remedy".
* To simplify things I suggest to have only the status "violated": if a Prohibition has no remedy this is the final status of the rule, if there is a remedy it has to be fulfilled: if it has been fulfilled the final status of the Prohibition is "ok", if not the status is still "violated".
* what does "(Not-)Confirmed" mean? By my understanding the term "confirm(ing)" implies taking an action but ODRL raises not need to tell the assigner "we confirm the prohibition Prohib997". Maybe "(Not-)Accepted" is a better term, or "(Not-)Complying-With"

**Comment by riannella** [comment]
> The handling of an internal infringed/violated status is inconsistent for the different properties of type Duty.

I would state this more as the "handling of statuses is dependent on the semantics of the different properties"

> re "infringed": it is a surprised to see "infringed" as status of Permission

I added this row as, otherwise, and ODRL evaluator would not know when (or if) a consequence is required to be fulfilled. (?)

> I think a row intentionally without a status in the "action" column doesn't make sense. Or is only the "Not-Exercised" missing?

Since the constraint is "not-satisfied", then it does not matter what the action cell says?

> For a full coverage of stati the case Infringed/Satisfied/ / Fulfilled/ /Not-Exercised/Not-Fulfilled should be included.

Yes, what others states are missing :-)

[states.pdf](https://github.com/w3c/poe/files/1247402/states.pdf)

**Comment by nitmws** [comment]
> I added this row as, otherwise, and ODRL evaluator would not know when (or if) a consequence is required to be fulfilled. (?)

Sorry, but consequence is a property of a duty Duty and not of a Permission. If a Permission has 3 duty-ies, each with a consequence,  2 of them are fulfilled and 1 of them is infringed, then only the consequence of this infringed duty has to be executed. So the Permission has not an "infringed" state and at the level of a permission no consequence can be executed as there is no such property.

> Since the constraint is "not-satisfied", then it does not matter what the action cell says?

If an ODRL processor should make use of this table each column should be filled in. If a basic rule for processing defines the action must be included in the processing we should explicitly say "in this case the execution of the action is '_not relevant_' " (Personal note: I've been struggling with such undefined states in other technical rules/standard and don't like that.)
And 1: I think we should narrow down the semantics of this column an head it "action exercised". The processor does not have to check if there as an action defined, it has only to check: has the action been exercised and this could be a Yes or No or a "not relevant"

And 2: if a column cannot be filled in for a Rule sub-class for formal reasons we should apply a NA (not applicable): remedy column in Permission and Duty rows, duty and consequence column in Prohibition rows. 

And 3: This table needs a legend for a full understanding: 
* multiple duty instances in a Permission and multiple remedy instances in a Prohibition may occur. In this case they must be ANDed
* consequence is not a property of Permission but of a duty.
* a definition of all terms used in the cells, or at least a reference to a section of the IM where all of them are defined.

> ... what others states are missing

I'll go over that ...

**Comment by nitmws** [comment]
I'm going over it and created a temporary Google Sheet as working document - the URL has been sent to all Members of the POE WG.

**Comment by riannella** [comment]
I have updated the IM to (hopefully) reflect the changes at the Thursday call.
Added to the model is the "refinement" property for Action/Asset/Party/Collection Constraints.
(constraint property is now only for Rules)

I have updated the Constraint section with subsections for refinement and constraint properties.

I have removed all mention of "effective" and "in effect"

And added a new section "Active Rule Processing" where we talk about how a rule is deemed Active. (I hope I got that right!)

(I have not updated the Vocab)
